support/refc/runtime.c

```diff
@@ -15,25 +15,78 @@ void push_Arglist(Value_Arglist *arglist, Value *arg) {
   arglist->filled++;
 }
 
+int isUnique(Value *value) {
+  if (value) {
+    return value->header.refCounter == 1;
+  }
+  return 0;
+}
+
+// necessary because not every variable passed as arguments is duplicated
+void removeArglistWithoutArgs(Value_Arglist *arglist) {
+  IDRIS2_REFC_VERIFY(arglist->header.refCounter > 0, "refCounter %lld",
+                     (long long)arglist->header.refCounter);
+  arglist->header.refCounter--;
+  if (arglist->header.refCounter == 0) {
+    free(arglist->args);
+    free(arglist);
+  }
+}
+
+void removeClosureWithoutArgs(Value_Closure *clos) {
+  IDRIS2_REFC_VERIFY(clos->header.refCounter > 0, "refCounter %lld",
+                     (long long)clos->header.refCounter);
+  clos->header.refCounter--;
+  if (clos->header.refCounter == 0) {
+    removeArglistWithoutArgs(clos->arglist);
+    free(clos);
+  }
+}
+
+void removeReuseConstructor(Value_Constructor *constr) {
+  if (!constr) {
+    return;
+  }
+  IDRIS2_REFC_VERIFY(constr->header.refCounter > 0, "refCounter %lld",
+                     (long long)constr->header.refCounter);
+  constr->header.refCounter--;
+  if (constr->header.refCounter == 0) {
+    if (constr->name) {
+      free(constr->name);
+    }
+    free(constr->args);
+    free(constr);
+  }
+}
+
 Value *apply_closure(Value *_clos, Value *arg) {
   // create a new arg list
   Value_Arglist *oldArgs = ((Value_Closure *)_clos)->arglist;
   Value_Arglist *newArgs = newArglist(0, oldArgs->total);
   newArgs->filled = oldArgs->filled + 1;
   // add argument to new arglist
   for (int i = 0; i < oldArgs->filled; i++) {
-    newArgs->args[i] = newReference(oldArgs->args[i]);
+    /*
+    if the closure has multiple references, then apply newReference to arguments
+    to avoid premature clearing of arguments
+    */
```

Comment:
You use this pattern a couple of times. Should it be a utility function?

How do we know that incrementing the reference wont lead to memory leaks?

Reply:
Each increment the reference will be interrupted by a decrement after the method execution `f(newArgs)`.